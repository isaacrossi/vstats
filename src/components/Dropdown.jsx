import { useEffect, useRef, useState } from "react";
import useOutsideClick from "../hooks/useOutsideClick";
import { GoChevronDown } from "react-icons/go";

const Dropdown = ({
  id,
  title = "select",
  data,
  hasImage,
  selectedId,
  onSelect,
}) => {
  const [isOpen, setIsopen] = useState(false);
  const [selectedItem, setSelectedItem] = useState(
    selectedId ? data?.find((item) => item.id === selectedId) : undefined
  );

  const handleChange = (item) => {
    setSelectedItem(item); //update the selectedItem state with the selected item
    onSelect && onSelect(item.id); // Call onSelect prop function with the selected item's id/ *this is a shorthand for if(onSelect) { onSelect(item.id) }
    setIsopen(false); //close the dropdown
  };

  useEffect(() => {
    if (selectedId && data) {
      const newSelectedItem = data.find((item) => item.id === selectedId);
      newSelectedItem && setSelectedItem(newSelectedItem);
    } else {
      setSelectedItem(undefined);
    }
  }, [selectedId, data]);

  const dropdownRef = useRef(null);
  useOutsideClick({
    ref: dropdownRef,
    handler: () => setIsopen(false),
  });

  return (
    <div ref={dropdownRef} className="relative">
      <button
        id={id}
        aria-label="Toggle dropdown"
        aria-haspopup="true"
        aria-expanded={isOpen}
        type="button"
        onClick={() => setIsopen(!isOpen)}
        className="flex align-center justify-between w-64 py-2 bg-transparent border-b border-red-600 text-slate-50"
      >
        <span>{selectedItem?.name || title}</span>
        <GoChevronDown
          size={20}
          className={`transform duration-500 ease-in-out text-red-600 ${
            isOpen ? "rotate-180" : ""
          }`}
        />
      </button>
      {isOpen && (
        <div className="absolute bg-blue-950 mt-2 w-max max-h-52 overflow-y-auto py-3 shadow-md rounded-md z-10">
          <ul
            className="bg-blue-950 w-64"
            role="menu"
            aria-labelledby={id}
            aria-orientation="vertical"
          >
            {data.map((item) => (
              <li
                className="flex items-center px-4 py-2 text-slate-50 hover:text-red-600 cursor-pointer"
                key={item.id}
                onClick={() => handleChange(item)}
              >
                {hasImage && (
                  <img
                    className="mr-2"
                    src={item.imageUrl}
                    alt="team"
                    loading="lazy"
                  />
                )}
                <span>{item.name}</span>
              </li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};

export default Dropdown;
