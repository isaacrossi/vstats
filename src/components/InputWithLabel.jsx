export const InputWithLabel = ({
  id,
  value,
  type = "text",
  onInputChange,
  onSubmit,
  isFocused,
  children,
}) => (
  <form onSubmit={onSubmit} className="mb-6">
    <label htmlFor={id} className="sr-only">
      {children}
    </label>
    &nbsp;
    <input
      id={id}
      type={type}
      value={value}
      autoFocus={isFocused}
      onChange={onInputChange}
      autoComplete="off"
      placeholder="Search for a player"
      className="w-64 py-2 bg-transparent border-b border-red-600 bg-search-icon bg-no-repeat bg-right bg-auto text-slate-50"
    />
  </form>
);
