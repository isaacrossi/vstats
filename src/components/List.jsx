import { shortenTeamName } from "../utils/shortenTeamName";
import { countries } from "../data/countries";

export const List = ({ list }) => (
  <ul>
    {list.map((item) => (
      <Item key={item.player.id} item={item} />
    ))}
  </ul>
);

const Item = ({ item }) => (
  <li className="grid grid-cols-4 text-s uppercase text-slate-50 px-5 py-4 border-b border-red-600">
    <span>{item.player.name}</span>
    <span className="flex">
      <img
        className="w-6 h-6 mr-2"
        src={item.statistics[0].team.logo}
        alt={item.statistics[0].team.name + "logo"}
      />
      {shortenTeamName(item.statistics[0].team.name)}
    </span>
    <span className="flex">
      <img
        className="h-5 w-auto mr-2"
        src={`https://flagsapi.com/${
          countries[item.player.nationality]
        }/flat/64.png`}
        alt="country flag"
      />
      {item.player.nationality}
    </span>
    <span>{item.statistics[0].games.position}</span>
  </li>
);
