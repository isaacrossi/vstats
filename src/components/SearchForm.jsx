import { GoX } from "react-icons/go";
import { GoSearch } from "react-icons/go";

const SearchForm = ({
  searchTerm,
  submittedSearchTerm,
  onSearchInput,
  onSearchSubmit,
  onSearchCancel,
}) => (
  <form onSubmit={onSearchSubmit} className="mb-6 relative max-w-80 h-auto">
    <InputWithLabel
      id="search"
      label="Search"
      value={searchTerm}
      onInputChange={onSearchInput}
    >
      Search
    </InputWithLabel>
    {submittedSearchTerm ? (
      <CancelButton onClick={onSearchCancel} />
    ) : (
      <SearchButton searchTerm={searchTerm} />
    )}
  </form>
);

const InputWithLabel = ({
  id,
  value,
  type = "text",
  onInputChange,
  isFocused,
  children,
}) => (
  <div>
    <label htmlFor={id} className="sr-only">
      {children}
    </label>
    &nbsp;
    <input
      id={id}
      type={type}
      value={value}
      autoFocus={isFocused}
      onChange={onInputChange}
      autoComplete="off"
      placeholder="Search must be at least 4 characters"
      className="w-80 py-2 bg-transparent border-b border-red-600 text-slate-50"
    />
  </div>
);

const SearchButton = ({ searchTerm }) => (
  <button
    type="submit"
    aria-label="Submit search"
    disabled={searchTerm.length > 0 && searchTerm.length < 4}
    className="absolute right-0 bottom-3"
  >
    <GoSearch size={20} className="text-red-600" />
  </button>
);

const CancelButton = ({ onClick }) => (
  <button
    type="button"
    aria-label="Cancel search"
    className="absolute right-0 bottom-3"
    onClick={onClick}
  >
    <GoX size={20} className="text-red-600" />
  </button>
);

export { SearchForm, InputWithLabel };
