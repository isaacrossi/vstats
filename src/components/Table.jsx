import { shortenTeamName } from "../utils/shortenTeamName";
import { countries } from "../data/countries";

const headers = ["Player", "Team", "Country", "Role"];

// export const List = ({ list }) => (
//   <ul>
//     {list.map((item) => (
//       <Item key={item.player.id} item={item} />
//     ))}
//   </ul>
// );

// const Item = ({ item }) => (
//   <li className="grid grid-cols-4 text-s uppercase text-slate-50 px-5 py-4 border-b border-red-600">
//     <span>{item.player.name}</span>
//     <span className="flex">
//       <img
//         className="w-6 h-6 mr-2"
//         src={item.statistics[0].team.logo}
//         alt={item.statistics[0].team.name + "logo"}
//       />
//       {shortenTeamName(item.statistics[0].team.name)}
//     </span>
//     <span className="flex">
//       <img
//         className="h-5 w-auto mr-2"
//         src={`https://flagsapi.com/${
//           countries[item.player.nationality]
//         }/flat/64.png`}
//         alt="country flag"
//       />
//       {item.player.nationality}
//     </span>
//     <span>{item.statistics[0].games.position}</span>
//   </li>
// );

export const Table = ({ list }) => (
  <table className="mt-14 mx-auto">
    <thead>
      <tr className="text-left grid grid-cols-4 gap-x-32 p-5 text-s uppercase text-slate-500 border-b border-red-600">
        {headers.map((header, index) => (
          <th key={index}>{header}</th>
        ))}
      </tr>
    </thead>
    <tbody>
      {list.map((item) => (
        <tr
          key={item.player.id}
          className="grid grid-cols-4 gap-x-32 p-5 text-s uppercase text-slate-50 border-b border-red-600"
        >
          <td className="w-56">{item.player.name}</td>
          <td className="flex w-56">
            <img
              className="w-6 h-6 mr-2"
              src={item.statistics[0].team.logo}
              alt={item.statistics[0].team.name + "logo"}
            />
            {shortenTeamName(item.statistics[0].team.name)}
          </td>
          <td className="flex w-56">
            <img
              className="h-5 w-auto mr-2"
              src={`https://flagsapi.com/${
                countries[item.player.nationality]
              }/flat/64.png`}
              alt="country flag"
            />
            {item.player.nationality}
          </td>
          <td className="w-56">{item.statistics[0].games.position}</td>
        </tr>
      ))}
    </tbody>
  </table>
);
