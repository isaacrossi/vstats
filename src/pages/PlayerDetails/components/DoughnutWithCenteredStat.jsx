import { Doughnut } from "react-chartjs-2";
import { Chart } from "react-chartjs-2";
import "chart.js/auto";

export const DoughnutWithCenteredStat = ({
  doughnutData,
  colourOne,
  colourTwo,
  centeredData,
  statText,
  isDark = false,
}) => (
  <div className="w-48 mx-auto relative mb-6 md:mb-9 ">
    <Doughnut
      data={{
        datasets: [
          {
            data: doughnutData,
            backgroundColor: [colourOne, colourTwo],
            cutout: "80%",
            borderWidth: 0,
          },
        ],
      }}
    />
    <div
      className={`absolute inset-0 flex flex-col items-center justify-center text-center mt-6 ${
        isDark ? "text-slate-50" : "text-slate-900"
      }`}
    >
      <h4 className="text-xl lg:text-2xl font-heading text-center pt-5 mb-1">
        {centeredData}
      </h4>
      <p className="uppercase text-sm lg:text-base text-center pb-5">
        <p className="uppercase text-sm lg:text-base text-center pb-5">
          {(statText ?? "").includes(" ")
            ? statText.split(" ").map((word, index) => (
                <span key={index} className="block">
                  {word}
                </span>
              ))
            : statText}
        </p>
      </p>
    </div>
  </div>
);
