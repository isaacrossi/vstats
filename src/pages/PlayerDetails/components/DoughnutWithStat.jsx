import { Doughnut } from "react-chartjs-2";
import { Chart } from "react-chartjs-2";
import "chart.js/auto";

export const DoughnutWithStat = ({
  doughnutData,
  colourOne,
  colourTwo,
  centeredData,
  statText,
}) => (
  <div className="w-48 mx-auto relative mb-6 md:mb-9">
    <Doughnut
      data={{
        datasets: [
          {
            data: doughnutData,
            backgroundColor: [colourOne, colourTwo],
            cutout: "80%",
          },
        ],
      }}
    />
    <div className="absolute inset-0 flex flex-col items-center justify-center text-center">
      <h4 className="text-xl lg:text-2xl font-heading text-center pt-5 mb-1">
        {centeredData}
      </h4>
      <p className="uppercase text-sm lg:text-base text-center pb-5">
        {statText}
      </p>
    </div>
  </div>
);
