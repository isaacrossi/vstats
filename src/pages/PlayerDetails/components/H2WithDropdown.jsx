import { H2WithSlash } from "./H2WithSlash";
import { Dropdown } from "../../../shared/components/Dropdown";

export const H2WithDropdown = ({
  dropdownData,
  handleDropdownChange,
  selectedItemId,
}) => (
  <div className="flex flex-col md:flex-row md:items-center justify-between w-full mb-4 md:mb-6">
    <H2WithSlash title="General" marginBottom={false} hasinput={true} />
    <Dropdown
      data={dropdownData}
      title="2024"
      id="league-dropdown"
      hasImage={false}
      onChange={handleDropdownChange}
      selectedItemId={selectedItemId}
    />
  </div>
);
