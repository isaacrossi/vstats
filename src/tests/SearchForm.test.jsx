import { describe, it, expect, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import { SearchForm, InputWithLabel } from "../components/SearchForm";

describe("SearchForm", () => {
  const searchFormProps = {
    searchTerm: "test",
    submittedSearchTerm: "",
    onSearchInput: vi.fn(),
    onSearchSubmit: vi.fn(),
    onSearchCancel: vi.fn(),
  };

  it("renders a clickable search button", () => {
    render(<SearchForm {...searchFormProps} />);
    expect(
      screen.getByRole("button", { name: "Submit search" })
    ).toBeInTheDocument();
  });

  it("does not render a cancel button when no search term has been submitted", () => {
    render(<SearchForm {...searchFormProps} />);
    expect(
      screen.queryByRole("button", { name: "Cancel search" })
    ).not.toBeInTheDocument();
  });

  it("calls onSearchInput when the input field value changes", () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.change(screen.getByDisplayValue("test"), {
      target: { value: "new value" },
    });

    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
  });

  it("calls onSearchSubmit when the SearchButton is clicked", () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.click(screen.getByRole("button", { name: "Submit search" }));
    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  });

  it("calls onSearchCancel when the CancelButton is clicked", () => {
    render(<SearchForm {...searchFormProps} submittedSearchTerm="thomas" />);
    fireEvent.click(screen.getByRole("button", { name: "Cancel search" }));
    expect(searchFormProps.onSearchCancel).toHaveBeenCalledTimes(1);
  });

  it("disables the search button when the search term length is between 1 and 3 characters", () => {
    render(<SearchForm {...searchFormProps} searchTerm="tho" />);
    expect(
      screen.getByRole("button", { name: "Submit search" })
    ).toBeDisabled();
  });
});

describe("InputWithLabel", () => {
  const inputProps = {
    id: "search",
    value: "test",
    children: "Search",
    onInputChange: vi.fn(),
    type: "text",
    isFocused: true,
  };

  it("renders the input field with its value", () => {
    render(<InputWithLabel {...inputProps} />);
    expect(screen.getByDisplayValue("test")).toBeInTheDocument();
  });

  it("renders the correct label for the input field", () => {
    render(<InputWithLabel {...inputProps} />);
    expect(screen.getByLabelText("Search")).toBeInTheDocument();
  });

  it("renders the correct placeholder text", () => {
    render(<InputWithLabel {...inputProps} />);
    expect(
      screen.getByPlaceholderText("Search must be at least 4 characters")
    ).toBeInTheDocument();
  });

  it("renders the correct id for the input field", () => {
    render(<InputWithLabel {...inputProps} />);
    expect(screen.getByDisplayValue("test").id).toBe("search");
  });

  it("renders the correct type for the input field", () => {
    render(<InputWithLabel {...inputProps} />);
    expect(screen.getByDisplayValue("test").type).toBe("text");
  });

  it("renders the input field with autoFocus", () => {
    render(<InputWithLabel {...inputProps} />);
    expect(screen.getByDisplayValue("test")).toHaveFocus();
  });

  it("calls onInputChange when the input field value changes", () => {
    render(<InputWithLabel {...inputProps} />);
    fireEvent.change(screen.getByDisplayValue("test"), {
      target: { value: "new value" },
    });
    expect(inputProps.onInputChange).toHaveBeenCalledTimes(1);
  });
});
