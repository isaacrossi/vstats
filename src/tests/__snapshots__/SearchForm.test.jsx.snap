// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="mb-6 relative max-w-80 h-auto"
  role="form"
>
  <div>
    <label
      class="sr-only"
      for="search"
    >
      Search
    </label>
    Â 
    <input
      autocomplete="off"
      class="w-80 py-2 bg-transparent border-b border-red-600 text-slate-50"
      id="search"
      placeholder="Search must be at least 4 characters"
      type="text"
      value="test"
    />
  </div>
  <button
    aria-label="Submit search"
    class="absolute right-0 bottom-3"
    type="submit"
  >
    <svg
      class="text-red-600"
      fill="currentColor"
      height="20"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10.25 2a8.25 8.25 0 0 1 6.34 13.53l5.69 5.69a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215l-5.69-5.69A8.25 8.25 0 1 1 10.25 2ZM3.5 10.25a6.75 6.75 0 1 0 13.5 0 6.75 6.75 0 0 0-13.5 0Z"
      />
    </svg>
  </button>
</form>
`;
